### ISSUE TITLE
Channel doesn't work in Testflight.

### ISSUE THREAD

**Issue by Extollite** [apple/swift-nio#1671]
### Expected behavior
Once I bind socket there will be reads in channel.
### Actual behavior
No reads in channel.

### Steps to reproduce

1. Create iOS app with UDPEcho in it.
2. Archive it and distribute via test flight.
3. Install app
4. Open app, then close and kill it.
5. Uninstall app.
6. Install it again.
7. Run server in app.
8. No reads in channel.

### If possible, minimal yet complete reproducer code (or URL to code)

https://hasteb.in/agolucir.swift

### SwiftNIO version/commit hash

SwiftNIO 2.23.0

### Swift & OS version (output of `swift --version && uname -a`)
```
Apple Swift version 5.3 (swiftlang-1200.0.29.2 clang-1200.0.30.1)
Target: x86_64-apple-darwin19.6.0
Darwin iMac-Pro.local 19.6.0 Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; root:xnu-6153.141.2~1/RELEASE_X86_64 x86_64

iOS 14.0.1
```

**Comment by glbrntt** [comment]
What happens the second time you run the server? Does the server successfully bind or do you get an error?

**Comment by Extollite** [comment]
Yes, it does.
```
The operation couldn’t be completed. (NIO.IOError error 1.)
```

**Comment by glbrntt** [comment]
```swift
            printer.print(error.localizedDescription)
```

Can you print the `error` (i.e. not `error.localizedDescription`)?

**Comment by Extollite** [comment]
I get this after starting the server and then trying to run again.
```bind(descriptor:ptr:bytes:): Address already in use (errno: 48)```

**Comment by glbrntt** [comment]
I think this is expected: the socket is still bound to "0.0.0.0:9999" from the first instance of the app. That it works on the simulator and not a real device is a quirk of the simulator, I suspect.

**Comment by Extollite** [comment]
Error is not the issue, issue is that when I try to connect to this socket from other device there is no incoming packets. I can connect with nc -u IP 9999 to this socket, but I don't get any echo response.

**Comment by Lukasa** [comment]
The error _is_ the issue: if the bind fails, you will never get reads, as the connection has not occurred.

On your bootstrap set `.channelOption(.socketOption(.so_reuseaddr), 1)`. This should help.

**Comment by Lukasa** [comment]
Sorry, I see you set that already.

It is essentially impossible to see a way for an app that is _uninstalled_ and _reinstalled_ to lead to a port conflict. Can you arrange for the button to guarantee it only ever creates one server (that is, it prevents being pressed twice)? I see no way you can be having state from the previous run affect the new one if you are actually killing and uninstalling the app.

**Comment by Extollite** [comment]
I have that in code I send.
Maybe I'm not clear enough.
What I'm doing is distributing my app via TestFlight to my iPhone.
Once I open my app and click run button I don't get any errors, but when I try to connect to this socket from my computer with `nc -u IP 9999` I get connected but when I type something I don't get any echo response.

When I click run button again, without turning off the app, I got this error, which is believe how it should work.

**Comment by Extollite** [comment]
For some reason when I update my app with Testflight and run it, it works, but when I do fresh install my UDP server don't get any packets.

**Comment by Lukasa** [comment]
I'd just like to summarise my understanding so I know we're on the same page:

1. On real phones, the `bind` call succeeds.
2. Sometimes it sends echo packets back.
3. Sometimes it doesn't send _any_ packets back.

Is that right? Or does it _never_ get packets back.

**Comment by Extollite** [comment]
Never sends packet back. But the problem appears only in Testflight, because when I just use Xcode to run it on phone it works all right.

**Comment by Lukasa** [comment]
Are the IP addresses the same in both cases?

**Comment by Extollite** [comment]
Yes

**Comment by Lukasa** [comment]
Should they be? When plugged in to Xcode your phone has an extra IP address that belongs to the USB connection. Are you sure your computer and phone are on the same network, and is that IP address the one for that network?

**Comment by Extollite** [comment]
But they are still connected to same local network, and once again I don't have any issue while running from Xcode via USB. The issue appears only in Testflight, when they are still in same network.

**Comment by Lukasa** [comment]
Yes, I agree they're connected to the same local network. My question is whether you are using an address on that local network to try to connect, or whether you're using the network over USB.

**Comment by Extollite** [comment]
The local network address.

**Comment by Davidde94** [comment]
What about if you run the debug version of the app that was installed by Xcode, but *not* while the device is plugged into your Mac? The app should still run without a problem.

**Comment by Davidde94** [comment]
Another thing we could try, archiving and distributing over TestFlight will build your app in release mode. Let's try doing a release-mode build and installing via Xcode.

**Comment by Extollite** [comment]
In both cases it works, I will setup some local server and try to distribute it ad hoc.

**Comment by Extollite** [comment]
Ok, so I distributed it ad hoc in release mode and it is not working. Any idea why it stop to work after archiving?

**Comment by Davidde94** [comment]
Just to confirm, you've changed "Build and Run" inside Xcode to use your release configuration, instead of debug, and it still works when installing the app via a USB cable?

**Comment by Extollite** [comment]
![Zrzut ekranu 2020-10-13 o 12 49 12](https://user-images.githubusercontent.com/42713788/95851440-8385de80-0d52-11eb-98b7-a01fbed51a74.png)
Yes

**Comment by Extollite** [comment]
Ok I found out, that turning off `Rebuild from Bitcode` while distributing ad hoc fixes issue.

**Comment by Extollite** [comment]
Ok, so looks like setting **Build Settings > Build Options > Enable Bitcode** to **No** fixes issue. Still no idea why it is like that.

**Comment by Extollite** [comment]
And I was wrong, something bugged and now it is same all over again. After archiving it stops to work.

**Comment by Davidde94** [comment]
@Extollite your link directs to an empty file, any chance you have a basic repo up on Github that I could take a look at?

**Comment by Extollite** [comment]
Ok, I did this one fast for testing https://github.com/Extollite/TestUDP.
I still experience same issue.

**Comment by Davidde94** [comment]
@Extollite you're intentionally using `DatagramBootstrap` instead of `ServerBootstrap` right?

**Comment by Davidde94** [comment]
@Extollite thanks for the sample. I've built, archived, and distributed to my own phone *with* bitcode enabled, and the app works as expected...

I did notice something interesting however. The app didn't work until I had granted permission to discover devices over the local network. I wonder perhaps if this is turned off on your device? I appreciate that you've said that it sometimes works, but this is probably a good starting point.

In your example code I changed the IP address from "0.0.0.0" to the IP address my phone had at the time "192.168.1.x"
